rocks.ml

```diff
@@ -215,13 +224,232 @@ module rec Iterator : Rocks_intf.ITERATOR with type db := RocksDb.t = struct
     !@err_pointer
 end
 
+and Transaction : Rocks_intf.TRANSACTION with type db := RocksDb.t and type iter := Iterator.t = struct
+  module ReadOptions = Rocks_options.ReadOptions
+  module WriteOptions = Rocks_options.WriteOptions
+  module TransactionOptions = Rocks_options.TransactionOptions
+  module Snapshot = Rocks_options.Snapshot
+
+  let name = "transaction"
+  let destructor  = "rocksdb_" ^ name ^ "_destroy"
+
+  type db = t
+  let db = t
+
+  type nonrec t = t
+  let t = t
+
+  let txnbegin_raw =
+    foreign
+      "rocksdb_transaction_begin"
+      (db @-> WriteOptions.t @-> TransactionOptions.t @-> ptr void @->
+         returning t)
+
+  let destroy = make_destroy t destructor
+
+  let txnbegin ?wopts ?txnopts db =
+    let inner wopts txnopts =
+      let t = txnbegin_raw db wopts txnopts null in
+      Gc.finalise destroy t;
```

Comment:
This finalizer in combination with the `with_t` below is probably the source of your double freeing problems ...